### PR TITLE
Add improvements to media service operations.

### PR DESCRIPTION
**Pull request by dewniMW**
This PR adds the following improvements:

- Add the capability to override the properties defined in the media.properties file via environment variables.
- Introduce a configuration to define the maximum allowed byte size of the media to be uploaded. The default value would 512 KB.
- Have the 'yyyyMMdd' date format (of the media uploaded date) appended to the unique id of the media and incorporate this when creating the hierarchical folder structure.
- Enforce user to manually create a directory called 'media' in the media mount location. This decision was taken as an additional safety measure to prevent files from being stored in an incorrect location. 
- Have user id instead of the username.
- Add ETag support for media download API.

### CODE REVIEW

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 1]
```diff
@@ -274,17 +290,111 @@ public void validateFileUploadMediaTypes(String mediaTypePathParam, String conte
...
     public void validateMediaTypePathParam(String mediaTypePathParam) throws StorageSystemClientException {
 
-        HashMap<String, List<String>> allowedContentTypes = StorageSystemUtil.getContentTypes();
-        if (MapUtils.isEmpty(allowedContentTypes) || !allowedContentTypes.keySet().contains(mediaTypePathParam)) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Started validating if the media content type path parameter in the request is a " +
+                    "supported content type.");
+        }
+
+        String environmentVariableForAllowedContentTypes = System.getenv(CONFIGURABLE_MEDIA_CONTENT_TYPES);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:298]: the variable name seems too long.. can we make it short?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java [hunk 2]
```diff
@@ -79,7 +78,7 @@
 
     @Override
     public String addMedia(List<InputStream> inputStreams, MediaMetadata mediaMetadata, String uuid,
-                           String tenantDomain) throws StorageSystemException {
+                           String tenantDomain) throws StorageSystemServerException {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java:81]: why was this changed? don't we need to send a client exception in uuid already exists?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java [hunk 3]
```diff
@@ -90,11 +89,11 @@ public String addMedia(List<InputStream> inputStreams, MediaMetadata mediaMetada
         } catch (IOException e) {
             throw new StorageSystemServerException("Error while uploading media to file system.", e);
         }
-
     }
 
     @Override
-    public DataContent getFile(String id, String tenantDomain, String type) throws StorageSystemException {
+    public DataContent getFile(String id, String tenantDomain, String type) throws StorageSystemServerException,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java:95]: this is not needed.. you can return the base exception class and then from the API component do an instance check..

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java (hunk 2 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java:81]: uuid is system generated. hence there is no check for uuid already exists condition.

File: components/org.wso2.carbon.identity.media.core/pom.xml [hunk 4]
```diff
@@ -108,23 +108,25 @@
...
+                            org.apache.commons.collections;version="${org.apache.commons.collections.imp.pkg.version.range}",
                             org.apache.commons.io;version="${org.apache.commons.io.imp.pkg.version.range}",
-                            org.wso2.carbon.identity.core.*;version="${carbon.identity.framework.imp.pkg.version.range}",
-                            org.apache.commons.codec.digest;version="${org.apache.commons.codec.imp.pkg.version.range}",
                             org.apache.commons.lang;version="${org.apache.commons.lang.imp.pkg.version.range}",
                             org.apache.commons.logging;version="${org.apache.commons.logging.imp.pkg.version.range}",
-                            org.apache.commons.collections;version="${org.apache.commons.collections.imp.pkg.version.range}",
                             org.json.simple,
-                            org.json.simple.parser
+                            org.json.simple.parser,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/pom.xml:116]: shall we add a version range for this

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/FileContentImpl.java [hunk 5]
```diff
@@ -27,16 +27,18 @@
 
     private File file;
     private String responseContentType;
+    private String etag;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/FileContentImpl.java:30]: let's change the variable name to eTag

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 6]
```diff
@@ -100,77 +107,65 @@ public DataContent readContent(String id, String tenantDomain, String type) thro
...
-            return storageSystemFactory.getInstance().isDownloadAllowedForProtectedMedia(id, type, tenantDomain);
-        }
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("StorageSystemFactory object is null hence returning security evaluation result as false.");
+                            "domain %s.", type, mediaId, tenantDomain));
         }
-        return false;
+        StorageSystemFactory storageSystemFactory = getStorageSystemFactory(getMediaStoreType());
+        String userId = getUserIdFromUserName(username);
+        return storageSystemFactory.getInstance().isDownloadAllowedForProtectedMedia(mediaId, type, tenantDomain,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:143]: ideally getStorageSystemFactory should return an instance of a StorageSystemFactory. you need to call isDownloadAllowedForProtectedMedia on that instance. Therefore no need to call getInstance() on an instance (it getInstance is used with singleton class)

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 7]
```diff
@@ -253,16 +242,38 @@ public InputStream transform(String id, String type, String tenantDomain, InputS
...
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Started content type validation for media to be uploaded.");
+        }
+
+        String envForAllowedContentTypes = System.getenv(CONFIGURABLE_MEDIA_CONTENT_TYPES);
+        HashMap<String, List<String>> allowedContentTypes = new HashMap<>();
+        if (StringUtils.isNotBlank(envForAllowedContentTypes)) {
+            String[] contentTypes = envForAllowedContentTypes.split(",");
+            for (String contentType : contentTypes) {
+                String environmentVariableForAllowedContentSubTypes =
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:254]: variable name is too long

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 8]
```diff
@@ -274,17 +285,109 @@ public void validateFileUploadMediaTypes(String mediaTypePathParam, String conte
...
+    public String getUserIdFromUserName(String username) throws StorageSystemServerException {
 
-        return MediaServiceDataHolder.getInstance().getStorageSystemFactories().get(storageType);
+        try {
+            RealmService realmService = MediaServiceDataHolder.getInstance().getRealmService();
+            int tenantID = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantID);
+            if (userRealm != null) {
+                UserStoreManager userStoreManager = (UserStoreManager) userRealm.getUserStoreManager();
+                String userIdFromUserName = ((AbstractUserStoreManager) userStoreManager)
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:324]: does this work correctly for a secondary userstore user?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java [hunk 9]
```diff
@@ -353,31 +361,30 @@ private boolean isFileAccessAllowed(File file) throws IOException, ParseExceptio
...
+    private boolean isMediaManagementAllowed(File file, String userId) throws IOException, ParseException {
 
         try (InputStreamReader reader = new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8)) {
             JSONParser jsonParser = new JSONParser();
             Object metadata = jsonParser.parse(reader);
-            String resourceOwner = (String) ((JSONObject) metadata).get(MEDIA_RESOURCE_OWNER);
-            return StringUtils.isNotBlank(resourceOwner) &&
-                    resourceOwner.equals(PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername());
+            String resourceOwner = (String) ((JSONObject) metadata).get(MEDIA_RESOURCE_OWNER_ID);
+            return StringUtils.isNotBlank(resourceOwner) && resourceOwner.equals(userId);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java:376]: can use StringUtils.equals

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java [hunk 10]
```diff
@@ -497,30 +504,55 @@ private Path getStorageDirectory(String fileType, int tenantId, String id) {
...
+    private Path getMediaFolderPreCreatedLocation() throws StorageSystemServerException {
+
+        Path fileStorageLocation = null;
+        Path configurableRootFolder = getMediaMountBaseDirectory();
 
         if (configurableRootFolder != null) {
-            fileStorageLocation = configurableRootFolder.resolve(Paths.get(MEDIA_STORE + fileType));
+            fileStorageLocation = configurableRootFolder.resolve(Paths.get(PRE_CREATED_MEDIA_FOLDER));
+
+            // The system expects a pre-created folder called 'media' to exist in the configured media mount base
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/file/FileBasedStorageSystemImpl.java:523]: use multi-line comments

File: components/org.wso2.carbon.identity.media.core/pom.xml (hunk 4 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/pom.xml:116]: unable to provide a version range here. same approach has been followed at https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.dcr/pom.xml#L134

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java (hunk 8 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:320]: yes I have tested the flow for a secondary userstore user

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java (hunk 6 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java]: fixed with 93a57530615d50896104e8d6407e96a73f289b57

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 11]
```diff
@@ -274,17 +281,109 @@ public void validateFileUploadMediaTypes(String mediaTypePathParam, String conte
...
+
+    private StorageSystem getStorageSystem(String storageType) throws StorageSystemServerException {
+
+        StorageSystemFactory storageSystemFactory =
+                MediaServiceDataHolder.getInstance().getStorageSystemFactories().get(storageType);
+        if (storageSystemFactory == null) {
+            throw new StorageSystemServerException(String.format("Unable to obtain StorageSystemFactory for " +
+                    "configured media store type: %s", storageType));
+        }
+        return storageSystemFactory.getInstance();
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:386]: here also since storageSystemFactory is an instance variable, its not correct to call getInstance() on that. can we rename the getInstace method to getStorageSystem or something meaningful?

Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java]: fixed

**Review by emswbandara — APPROVED** [review]

**Review by ashensw — APPROVED** [review]
